qiskit/circuit/__init__.py

```diff
@@ -54,81 +78,825 @@
+equal if they are the same Python object, or if they were both created by a register of the same
+name and size, and they refer to the same index within that register.  There is also a special type
+tag for "ancilla" qubits, but this is little used in the current state
+of Qiskit:
+
+.. autoclass:: AncillaQubit
+    :show-inheritance:
+    :class-doc-from: class
+
+A collection bits of the same type can be encapsulated in a register of the matching type.  The base
+functionality is in a base class that is not directly instantiated:
+
+.. autoclass:: Register
+    :members:
+
+Each of the defined bit subtypes has an associated register, which have the same constructor
+signatures, methods and properties as the base class:
+
+.. autoclass:: QuantumRegister
+    :show-inheritance:
+    :class-doc-from: class
+.. autoclass:: ClassicalRegister
+    :show-inheritance:
+    :class-doc-from: class
+.. autoclass:: AncillaRegister
+    :show-inheritance:
+    :class-doc-from: class
+
+A common way to instantiate several bits at once is to create a register, such as by
+``QuantumRegister("my_qreg", 5)``.  This has the advantage that you can give that collection of bits
+a name, which will appear during circuit visualizations (:meth:`QuantumCircuit.draw`) and exports to
+interchange languages (see :mod:`.qasm2` and :mod:`.qasm3`).  You can also pass a name and a list of
+pre-constructed bits, but this creates an "aliasing register", which are very poorly supported on
+hardware.
+
+Circuits track registers, but registers themselves impart almost no behavioral differences on
+circuits.  The only exception is that :class:`ClassicalRegister`\ s can be implicitly cast to
+unsigned integers for use in conditional comparisons of :ref:`control flow operations
+<circuit-control-flow>`.
+
+Classical registers and bits were the original way of representing classical data in Qiskit, and
+remain the most supported currently.  Longer term, the data model is moving towards a more complete
+and strongly typed representation of a range of classical data (see
+:ref:`circuit-repr-near-time-classical`), but you will still very commonly use classical bits in
+current Qiskit.
+
+
+Instruction contexts
+--------------------
+
+The scalar type of the :attr:`QuantumCircuit.data` sequence is the "instruction context" object,
+:class:`CircuitInstruction`.  This is essentially just a data class that contains a representation
+of what is to be done (its :attr:`~CircuitInstruction.operation`), and the data it acts on (the
+:attr:`~CircuitInstruction.qubits` and :attr:`~CircuitInstruction.clbits`).
+
+.. autoclass:: CircuitInstruction
+    :members:
+
+Programmatically, this class is actually implemented in Rust and is a constructed handle to internal
+data within Rust space.  Mutations to instances of this class will not be reflected in the circuit.
+In general, you cannot mutate instruction contexts that are already in the circuit directly; the
```

Comment:
What about
`qc.data = [instr for instr in qc.data if instr.operation.num_qubits > 2]` as suggested in https://www.youtube.com/watch?v=cC0dFDv7XMQ ?
If we don't want user to do things like these we should probably not advertise them on youtube... :-)

Reply:
This isn't a mutation of a `CircuitInstruction` instance, it's modification of the container of instructions.

The Qiskit team don't have control over the youtube channel, but I would certainly not suggested that pattern had I been consulted.

Reply:
Alright, makes sense to me. :-)